File: lib/09.js
function parseInput(contents) {
	return contents.split(/\n/g).map(Number);
}

function checkValid(window, target) {
	for (let index = 0; index < window.length; index++) {
		const curr = window[index];
		const q = target - curr;
		if (window.slice(index + 1).includes(q)) {
			return true;
		}
	}
	return false;
}

function findInvalid(input, windowSize) {
	let position = windowSize;
	while (position < input.length) {
		const target = input[position];
		const check = input.slice(position - windowSize, position);

		if (!checkValid(check, target)) {
			return target;
		}

		position++;
	}
}

module.exports = {
	A09(contents, windowSize = 25) {
		const input = parseInput(contents);

		return findInvalid(input, windowSize);
	},

	B09(contents, windowSize = 25) {
		const input = parseInput(contents);

		const invalid = findInvalid(input, windowSize);

		let position = 0;
		while (position < input.length) {
			let numbers = [];
			let sum = 0;
			let count = 0;

			while (sum < invalid) {
				sum += input[position + count];
				numbers.push(input[position + count]);
				count++;
			}

			if (sum === invalid && count > 1) {
				return Math.min(...numbers) + Math.max(...numbers);
			}

			position++;
		}
	},
};

File: lib/09.dart
import 'dart:math';

List<int> parseInput(String contents) {
  return contents.split('\n').map((e) => int.parse(e)).toList();
}

bool checkValid(List<int> window, int target) {
  for (int index = 0; index < window.length; index++) {
    int curr = window.elementAt(index);
    int q = target - curr;
    if (window.sublist(index + 1).contains(q)) {
      return true;
    }
  }
  return false;
}

int findValid(List<int> input, int windowSize) {
  int position = windowSize;
  while(position < input.length) {
    int target = input.elementAt(position);
    List<int> check = input.sublist(position - windowSize, position);

    if (!checkValid(check, target)) {
      return target;
    }

    position++;
  }
}

int A09(String contents, {windowSize:25}) {
  List<int> input = parseInput(contents);
  
  return findValid(input, windowSize);
}

int B09(String contents, {windowSize:25}) {
  List<int> input = parseInput(contents);
  
  int invalid = findValid(input, windowSize);

  int position = 0;
  while (position < input.length) {
    List<int> numbers = [];
    int sum = 0;
    int count = 0;

    while(sum < invalid) {
      numbers.add(input.elementAt(position + count));
      sum += numbers.last;
      count++;
    }

    if (sum == invalid && count > 1) {
      return numbers.reduce(min) + numbers.reduce(max);
    }

    position++;
  }
}
